text_classification.py

```diff
@@ -0,0 +1,487 @@
+
+        Arguments:
+            split_ratio: split train_examples into train set (split_ratio)
+                and valid set (1-split_ratio). Default: 0.7
+            batch_size: batch size. Default: 32
+            device: the device to sent data. Default: 'cpu'
+            random_state: the random state provided by user. Default: None
+
+        Examples:
+            >>> train_iter, test_iter, valid_iter = txt_cls.iters(device="cpu")
+
+        Outputs:
+            - train_iter: a iterator based on
+                train.csv file with split_ratio percent.
+            - train_iter: a iterator based on test.csv file.
+            - valid_iter: a iterator based on
+                train.csv file with 1-split_ratio percent.
+
+        """
+
+        train, test, valid = self.split(split_ratio, random_state)
+        return generate_iterators(
+            (train, test, valid), [batch_size] * 3, device=device)
+
+    def download(self):
+        """Download the dataset if it doesn't exist in processed_folder already."""
+
+        train_csv_path = os.path.join(self.raw_folder,
+                                      self.dataset_name + '_csv',
+                                      'train.csv')
+        test_csv_path = os.path.join(self.raw_folder,
+                                     self.dataset_name + '_csv',
+                                     'test.csv')
+        if os.path.exists(train_csv_path) and os.path.exists(test_csv_path):
+            return
+
+        os.makedirs(self.raw_folder)
+        filename = self.dataset_name + '_csv.tar.gz'
+        url = self.url
+        path = os.path.join(self.raw_folder, filename)
+        download_from_url(url, path)
+        extract_archive(path, self.raw_folder, remove_finished=True)
+
+        print('Dataset %s downloaded.' % self.dataset_name)
+
+    def preprocess(self, raw_folder=None, processed_folder=None):
+        """Preprocess the csv files."""
+
+        if raw_folder is None:
+            raw_folder = os.path.join(self.raw_folder, self.dataset_name + '_csv')
+
+        if processed_folder is None:
+            processed_folder = self.processed_folder
+
+        if os.path.exists(processed_folder) is not True:
+            os.makedirs(processed_folder)
+
+        src_filepath = os.path.join(raw_folder, 'train.csv')
+        if not os.path.isfile(src_filepath):
+            self.download()
+        tgt_filepath = os.path.join(processed_folder, self.dataset_name + '.train')
+        self.text_normalize(src_filepath, tgt_filepath)
+
+        src_filepath = os.path.join(raw_folder, 'test.csv')
+        if not os.path.isfile(src_filepath):
+            self.download()
+        tgt_filepath = os.path.join(processed_folder, self.dataset_name + '.test')
+        self.text_normalize(src_filepath, tgt_filepath)
+
+        print("Dataset %s preprocessed." % self.dataset_name)
+
+    def text_normalize(self, src_filepath, tgt_filepath):
```

Comment:
Can we move this out of the class and make it a general function? This seems useful by itself. We just need to add a bit of an explanation I think (i.e. what kind of text this is useful for).

Reply:
Same as above